dask/bag/random.py

```diff
@@ -0,0 +1,123 @@
+import random
+
+
+def sample(population, k):
+    """
+    Chooses k unique random elements from a population bag.
+
+    Returns a new bag containing elements from the population while
+    leaving the original population unchanged.
+
+    Parameters
+    ----------
+    population : Bag
+        Elements to sample.
+    k : integer, optional
+        Number of elements to sample.
+    Examples
+    --------
+    >>> import dask.bag as db
+    >>> from dask.bag import random
+    >>>
+    >>> b = db.from_sequence(range(5), npartitions=2)
+    >>> len(list(random.sample(b, 3).compute()))
+    3
+    """
+    return _sample(population, k, False)
+
+
+def choices(population, k=1):
+    """
+    Return a k sized list of population elements chosen with replacement.
+
+    If the relative weights or cumulative weights are not specified,
+    the selections are made with equal probability.
+
+    Parameters
+    ----------
+    population : Bag
+        Elements to sample.
+    k : integer, optional
+        Number of elements to sample.
+    Examples
+    --------
+    >>> import dask.bag as db
+    >>> from dask.bag import random
+    >>>
+    >>> b = db.from_sequence(range(5), npartitions=2)
+    >>> len(list(random.choices(b, 3).compute()))
+    3
+    """
+    return _sample(population, k, True)
+
+
+def _sample(population, k=1, replace=False):
+    return population.map_partitions(_sample_map_partitions, k, replace).reduction(
+        lambda x: x, _sample_reduce
+    )
+
+
+def _sample_map_partitions(population, k=1, replace=True):
+    """
+    Map function used on the sample and choices functions.
+    Parameters
+    ----------
+    population : list
+        List of elements to sample.
+    k : int, optional
+        Number of elements to sample. Default is a single value is returned.
+    replace : boolean, optional
+        Whether the sample is with or without replacement
+
+    Returns a tuple, composed by:
+    - list of k samples;
+    - total number of elements from where the sample was drawn;
+    - total number of elements to be sampled;
+    - boolean which is True whether the sample is with or without replacement.
+    """
+    lx = len(population)
+    real_k = k if k <= lx else lx
+    sample_fun = random.choices if replace else random.sample
+    # because otherwise it raises IndexError:
+    sampled = [] if real_k == 0 else sample_fun(population, k=real_k)
```

Comment:
As you can see on this test unit, it can happen to have a partition with 0 elements, in this case real_k must be forced to zero, we shouldn't raise an exception on the maps since there can be other non-empty partitions, in other words, we shouldn't stop the entire sampling because of the existence of an empty partition :-)

```python
def test_choices_size_over_repartition():
    """
    Number of randomly sampled are more than the elements on each partition.
    """
    a = db.from_sequence(range(10), partition_size=9)
    a = a.repartition(3)
    s = random.choices(a, k=2)
    assert len(list(s.compute())) == 2
```


Regarding if we should raise an exception if k is 0 or the overall bag is empty, I checked the behaviour of random.choices function and it doesn't raise any. 
Here's the experiment:
``` python
>>> import random
>>> random.choices([1,2,3], k=0)
[]
>>> random.choices([], k=0)
[]
>>> 


```

Reply:
Ok. Thanks for the explanation. That makes sense to match the behavior of `random.choices`.